### ISSUE TITLE
Use custom container with alidock

### ISSUE THREAD

**Issue by mconcas** [alidock/alidock#145]
This is not really an issue, it is just to continue the thread on the alice-analysis-taskforce.
So, it is possible to use the another container with alidock.

In case it is published on docker-hub you can tell alidock to use it by creating a configuration file in your home called: `.alidock-config.yaml`

```
---
imageName: custom/container 
```
Custom containers can be derived directly using the alidock one as a base in order to have the whole facilities in place.
If you have a Dockerfile you can build your container, where possibly you install the needed software, then you can publish it on Docker Hub and use it with alidock. It works also with local containers not pushed on DH, but usually publishing it is convenient. 

To summarize, the hierarchy should look like:

| `alisw/slc7-builder` | `alidock/alidock`              | `custom/container`            |
|----------------------|--------------------------------|-------------------------------|
| _official image_   | *stuff to work with alidock* | *further custom requisites* |

**Comment by bovulpes** [comment]
To create a local new tag of the alidock image with stuff added by you do this:

> alidock root    (to enter the container as root)
> yum install ...  (to add your stuff) 
> exit   (exit the container, which will continue to run)
> docker  commit  -m "some message reminder"  - a "author"  alidock-1000  alisw/alidock:my_tag

first check with 

> docker ps

that the started container is indeed named "alidock-1000"

If you want to upload it on docker hub, if you already have an account there, create a new repository after logging on hub.docker.com let's say your-docker-id/your-alidock

Then, on you local machine first login with

> docker login docker.io

then tag the recently committed image with your docker ID

> docker tag alisw/alidock:my_tag your-docker-id/your-alidock:some-tag

and finally push your newly created image

> docker push your-docker-id/your-alidock:some-tag

and this is it. I hope it helps.

**Comment by mconcas** [comment]
Dear @bovulpes , you can also create a Dockerfile that includes all your additions, than you build it and push, without involving an interactive shell in alidock.

Dockerfile should look like:
```Dockerfile
FROM alipier/alidock

LABEL maintainer="User Name <username@cern.ch>"

RUN yum install -y ...your stuff
RUN rm -rf /var/cache/yum/*
```
then you build and tag it with:
```bash
docker build . -t repo/image
```

And then:
```bash
docker push your-docker-id/your-alidock:some-tag
```

more details here: [Dockerfile](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/)

**Comment by bovulpes** [comment]
Thanks @mconcas the more solutions the better! For larger modifications I also use your method.

**Comment by mconcas** [comment]
Ok, then I am closing this to but it still remains as a reference to the link posted in the mailing list.
Feel free to reopen this in case of further details/issues.

Cheers

**Comment by maireiphc** [comment]
Hi Matteo,

   sorry, I am lost. I do not know exactly what is finally the advised way to proceed to install extra packages (latex suite, in my case). I understand that I need to touch and tamper with the Dockerfile, but from where should I start ?

My knowledge starts/stops with the installation command : 
bash <(curl -fsSL https://raw.githubusercontent.com/alidock/alidock/master/alidock-installer.sh)
Héhé...

Cheers,
    Antonin

**Comment by mconcas** [comment]
No problem.
can you give me the command you use to install the missing packages?
I will write a Docker file you can use to build your container and as example for future edits.

Cheers

**Comment by maireiphc** [comment]
Hello Matteo,
It is taking longer than I thought to make it working. Currently facing font troubles once the first texlive packages are installed. Coming back to you here where the issue is solved to have a fully working example.

**Comment by maireiphc** [comment]
Hi Matteo,

  I have apparently a set of CentOS7 packages that seems to be working. 
I had issues to find relsize.sty from LaTeX; I needed to install it by hand at the end of the day.

I tried to keep the list minimalist, but ok, there are maybe redundancies.
It will be about (25+148 sub-dependency) packages to be installed  + 1 upgrade, = 135 MB

```
yum install \
texlive-base \
texlive-pstricks \
texlive-latex* \
texlive-collection-basic \
texlive-txfonts \
texlive-pslatex \
ps2eps \
texlive-dvips \
ImageMagick \
texlive-pst-tools \
ghostscript \
ghostscript-fonts \
texlive-mathpazo \
texlive-palatino texlive-cm texlive-cm-super texlive-times texlive-helvetic texlive-mnsymbol texlive-fpl

mkdir -p /usr/share/texlive/texmf-local/tex/latex/relsize
curl -o /usr/share/texlive/texmf-local/tex/latex/relsize/relsize.sty  https://ctan.crest.fr/tex-archive/macros/latex/contrib/relsize/relsize.sty
texhash
```


Cheers, Antonin

**Comment by mconcas** [comment]
Hi Antonin,
I created this dummy dockerfile to include your requirements.
I am not really sure if it can be avoided to be run as root, probably you can figure it out better.
To use it:

- create a file named Dockerfile and past the content
- run `docker build -t myrepo/myimage .`  where _myrepo_ and _myimage_ can be arbitrary names as long as you don't push them on Docker Hub [[1](https://docs.docker.com/docker-hub/repos/)]
- create a `.alidock-config.yaml` file in your home with these lines:
```yaml
---
imageName: myrepo/myimage:latest
```
- ensure other alidock containers are not running: `alidock stop`
- `alidock root` (this is to run `texhash` with no permission issues)

Here there is Dockerfile you can start from to fix all remaining permission issues:

```dockerfile
FROM alipier/alidock:latest

LABEL maintainer="User Name <username@userdomain.com>"
RUN yum install -y texlive-base                \
		   texlive-pstricks            \
		   texlive-latex*              \
		   texlive-collection-basic    \
		   texlive-txfonts             \
		   texlive-pslatex             \
		   ps2eps                      \
		   texlive-dvips               \
		   ImageMagick                 \
		   texlive-pst-tools           \
		   ghostscript                 \
		   ghostscript-fonts           \
		   texlive-mathpazo            \
		   texlive-palatino            \
		   texlive-cm                  \
		   texlive-cm-super            \
		   texlive-times               \
 		   texlive-helvetic            \
 		   texlive-mnsymbol            \
		   texlive-fpl             && \
    yum clean all && rm -rf /var/cache/yum

RUN mkdir -p /usr/share/texlive/texmf-local/tex/latex/relsize && \
    curl -o /usr/share/texlive/texmf-local/tex/latex/relsize/relsize.sty https://ctan.crest.fr/tex-archive/macros/latex/contrib/relsize/relsize.sty
```
Cheers,
Matteo

**Comment by mconcas** [comment]
PS. Notice that in case you want to play with rebuilding the image, in order to run the latest you have to run `alidock stop` every time, otherwise it will log into running container, which would be not the latest built.

Cheers

**Comment by maireiphc** [comment]
Hi Matteo,

  thank you for your prompt reply and for your details.
Still in trouble (my bad).

> docker build -t alidock/alidock+latex4rivet -f BookKeeping/Docker-Alidock/Dockerfile .
invalid argument "alidock/alidock+latex4rivet" for "-t, --tag" flag: invalid reference format

1. Where in the OS should/can I run the docker build command ?
(So far, I have /home/myuserID/alidock with the earlier alidock installation)

2. my tag "myrepo/myimage" is not accepted apparently. I haven't figured out why (no uppercase, removing the + does not help, ...)
I am certainly missing sthg (https://docs.docker.com/engine/reference/commandline/build/)

**Comment by mconcas** [comment]
> > docker build -t alidock/alidock+latex4rivet -f BookKeeping/Docker-Alidock/Dockerfile .
> > invalid argument "alidock/alidock+latex4rivet" for "-t, --tag" flag: invalid reference format

To me, if I remove the `+` it works:
`docker build -t alidock/alidocklatex4rivet -f Dockerfile`
  
> 1. Where in the OS should/can I run the docker build command ?

Wherever you want, if you are manually passing the path to Dockerfile like in your snippet. Docker != alidock, so it doesn't really matter.

> 2. my tag "myrepo/myimage" is not accepted apparently. I haven't figured out why (no uppercase, removing the + does not help, ...)

I suggest you not to use alidock/smth as tag since if at some point you would like to push it on docker hub this will not work since alidock repository exists already. Just try with something flat like: `amaire/alidocklatex4rivet`

Cheers,
Matteo

**Comment by maireiphc** [comment]
Ok, now it starts. Thanks !
It is expected to take long time + a certain disk space ?
"Sending build context to Docker daemon  55.17GB" and growing ...

**Comment by mconcas** [comment]
That sounds weird, mine takes less than 3GB with the latex package installed

**Comment by maireiphc** [comment]
Actually what the docker build is doing behind the scene ?
Taking a full CentOS 7 distro, add the layer alidock  and then run the yum install things that I wanted ?

**Comment by mconcas** [comment]
Depending on what is the base image you are using it downloads it.
In the alidock case it is used a base to start with your custom instructions (`FROM alipier/alidock`).
`alipier/alidock` base image size is 2.44GB large. 
Then you put on top some layers (they are called like that) of installed software.
Then the result is 2.8GB

**Comment by maireiphc** [comment]
. Should I have an alidock container running somewhere at the build time ? Or is it ok if I `alidock stop` before the `docker build` ?

. How Docker knows what it has to pick-up : https://github.com/alidock/alipier ?
`FROM alipier/alidock` , in the Dockerfile, is enough, it completes whatever is needed to locate the source ?

**Comment by mconcas** [comment]
> . Should I have an alidock container running somewhere at the build time ? Or is it ok if I `alidock stop` before the `docker build` ?

Running containers are not related to image building
> 
> . How Docker knows it is had to pick-up : https://github.com/alidock/alipier ?

it is not picking that from there, it is picking that from dockerhub, the `alipier/alidock` is a repository there: https://hub.docker.com/repository/docker/alipier/alidock

> `FROM alipier/alidock` , in the Dockerfile, is enough, it completes whatever is needed to locate the source ?

Yes

**Comment by maireiphc** [comment]
Ok, I stop the process while I have been reaching 150 GB of 'sending build context'
I did the `docker build` command directly under my home directory, which is apparently not recommended (https://docs.docker.com/engine/reference/commandline/build/ "In most cases, it’s best to put each Dockerfile in an empty directory" ... )

I created an empty directory (~HOME/alidock-latex4rivet/), moved there and `docker build` there.
The build worked (quickly).
After the `.alidock-config.yaml` created in my home dir, it seems that `alidock` is indeed launching the proper container and in there I find the necessary commands. (great !)

Now extra questions : 
. why you did _not_ include texhash in the dockerfile itself ?
. where is stored the built container itself  on my native arborescence on my laptop ?
. how can I remove it cleanly and exhaustively ? (`docker rm "myrepo/myimage"` ?)
. would it be possible to flip back on demand to the regular alidock container while keeping at hand the cursom one ?

In the dockerfile above, there is a typo with texlive-fpil  -> texlive-fpl, it means that I will have to rerun with the proper package installed with yum.

Cheers,
   Antonin

**Comment by mconcas** [comment]
> Ok, I stop the process while I have been reaching 150 GB of 'sending build context'
> I did the `docker build` command directly under my home directory, which is apparently not recommended (https://docs.docker.com/engine/reference/commandline/build/ "In most cases, it’s best to put each Dockerfile in an empty directory" ... )

True, i forgot that.

> . why you did _not_ include texhash in the dockerfile itself ?

why should I do that? isn't that the call to the program? this you can do in any session. Dockerfile just describes commands to setup environment and prereq. The entrypoint of alidock is its ssh server.

> . where is stored the built container itself on my native arborescence on my laptop ?

`$> docker images`
> . how can I remove it cleanly and exhaustively ? (`docker rm "myrepo/myimage"` ?)

`$> docker rmi myrepo/myimage` (rmi -> remove image)

> . would it be possible to flip back on demand to the regular alidock container while keeping at hand the cursom one ?

This is indeed an even more advantage usage, You surely can, but you should work with the master branch of alidock on github  and see from there how to use that (this is not a feature that I push people to do, since alidock is conceived for it's primary usage, which is quite limited by design).
> In the dockerfile above, there is a typo with texlive-fpil -> texlive-fpl, it means that I will have to rerun with the proper package installed with yum.

fixed.
> 
> Cheers,
> Antonin

Closing since topic switched to pure docker commands and other things not in the main question.
